[RFC 0165] Bootspec v2 (#165)
NixOS/rfcs: rfcs/0165-bootspecv2.md

Reviewer: We have a bunch of proposed nominations from @RaitoBezarius. I would appreciate if the nominees could accept or deny.

- @lheckemann 
- @GovanifY
- @06kellyjac (assumed typo, 👍 on the comment, please explicitly accept)
- @hesiod
- @amjoseph-nixpkgs

Just a reminder that there is no special qualification required to be a shepherd. The role just involves ensuring the RFC is moving forward by collecting feedback, ensuring the RFC text evolves to incorporate any raised concerns and eventually moving towards FCP. (The details of the process are documented in the [README](https://github.com/NixOS/rfcs/blob/master/README.md)).

Author: @lheckemann do you confirm to be co-author my dear linus? :p -- OK for shepherding, thank you!

Reviewer 2: Speaking as a member of the RFC steering committee, after people [confirmed their nominations](https://github.com/NixOS/rfcs/pull/165#issuecomment-1855690878), we approve of the shepherd team as is in the RFC:
- Leader: @JulienMalka
- Shepherds: @lheckemann (or co-author) @GovanifY @06kellyjac @hesiod

  (https://github.com/systemd/systemd/pull/28891) offers a more reliable and
  robust solution for generating signed PCR policies.
- Supporting SecureBoot.
  Secure Boot has one maintained implementation that is being upstreamed: https://github.com/nix-community/lanzaboote
  which was enabled by Bootspec v1.
- Specifying how to discover generations. This is desirable, but should not be tied to bootspec directly since bootspec may be useful with diverse discovery mechanisms.
- Address how bootloader backends should handle foreign bootloader data in their stead and manipulate it or how transition from a bootloader to another should happen.

# Proposed Solution
[proposed-solution]: #proposed-solution

- `initrd` will be removed from the v2
- `initrds` will be introduced as a list of initrd (compressed or uncompressed
  CPIO archives), this list can be empty, but the field is **required**
  nonetheless.
- `org.nixos.initrd-secrets.v1` will now be an **official extension** of
  Bootspec which should be an hashmap of strings (key) and strings (value)
  where the key is the "name" of a secret and the value is the **filesystem path** towards a
  secret.

> - `fdtdir` will be introduced as an opaque string to a directory in a shape as
>   the kernel outputs give them, e.g. U-Boot's extlinux directive `FDTDIR` is an
>   example of this behavior — it is **optional**

Reviewer 3: ```suggestion
- `fdtdir` will be introduced as an opaque string to a directory following the structure
   of the `dtbs` folder from the kernel, e.g. U-Boot's extlinux directive `FDTDIR` is an
  example of this behavior — it is **optional**
```

Alternatively to the `dtbs` folder, it could be phrased referencing the fact it's the `dtbs_install` target of the build.

---
feature: bootspec_v2
start-date: 2022-11-01
author: Ryan Lahfa (@raitobezarius)
co-authors: Linus Heckemann (@lheckemann)
shepherd-team:

> - @06kellyjack

Reviewer 4: ```suggestion
- @06kellyjac
```

- Multiple initrds support
- Rework of initrd secrets mechanism
- Lack of devicetree
- Further iterations of Bootspec

This document describes **Bootspec v2**.

# Motivation
[motivation]: #motivation

The motivation of Bootspec v1 remains but we address the v1 weaknesses and
includes our experience of running with it for a while.

## Multiple initrds

The Boot Loader Specification, developed by the Userspace API group, allows the
specification of multiple initrd entries [1], with the purpose of merging them into
CPIO archives. This functionality is valuable in the systemd-stub ecosystem,
where various types of initrds are combined, including specific credential
initrds, global credential initrds, system extension initrds, PCR signature
initrds, and PCR public key initrds. Some initrds, such as credential initrds,
are dynamically generated from an EFI System Partition (ESP) location. [2]

Moreover, an additional initrd is frequently used to store CPU microcode. To
ensure compatibility and flexibility, it is essential to rework the initrd
support in the Boot Loader Specification. The proposed changes aim to treat

> initrds as a set, allowing bootloaders to handle a list of initrds or a single

Author: That's correct but I guess I should add some remarks on what is possible in terms of ordering.
For example, CPU microcode must always come first, uncompressed.

> initrds as a set, allowing bootloaders to handle a list of initrds or a single

Reviewer 5: That makes it all the more important :D I was mostly thinking of precedence with colliding paths, but microcode-first is obviously especially important. I'm not sure we even need to detail it that much in the text, but we shouldn't call it a "set" :)

> initrds as a set, allowing bootloaders to handle a list of initrds or a single

Reviewer 5: ```suggestion
initrds as a list, allowing bootloaders to handle a list of initrds or a single
```

[2]: https://github.com/systemd/systemd/blob/main/src/boot/efi/stub.c#L778-L793

## Initrd secrets

Initrd secrets play a crucial role in the NixOS ecosystem, but they have raised
concerns, including security vulnerabilities (e.g., [GHSA-3rvf-24q2-24ww](https://github.com/NixOS/calamares-nixos-extensions/security/advisories/GHSA-3rvf-24q2-24ww
)) and
issues with booting in the Heads platform (e.g., [Issue #1348](https://github.com/linuxboot/heads/issues/1348.
)).

Initrd secrets serve to protect boot-time secrets from exposure within the Nix
store. This is achieved by using runtime scripts that append CPIO archives to
the generation's initrd during bootloader installation. However, the term
"initrd secrets" can be misleading, as the secrets are plaintext and can be
accessed from the ESP partition, offering limited confidentiality and no
integrity.

In practice, initrd secrets are often employed to establish stable fingerprints
for SSH servers within the initrd, or aiding in remote disk decryption on servers.
To address these issues, this RFC proposes moving away from the "appender

> script" model in Bootspec v1 and instead adopting a hash map format to

Author: Sure!

Regarding overlays in NixOS, any expressed overlay is incorporated into the final device tree. This eliminates the necessity to formalize an additional overlays field list, as overlays can be transformed into device trees as needed.

## Further iterations of Bootspec

It has been brought that the RFC process may not be adequate to discuss further
iterations of Bootspec.

While the author disagree with this vision and believe this would be a loss
for increased participation in the elaboration of further iterations and would lead to a 
decrease in discoverability of Bootspec matters.

We propose to move further iterations of Bootspec away from the RFC process.

For this, we propose to follow a lightweight process inside a GitHub repository
containing the specification and other relevant metadata about Bootspec.

In instances of controversy or difficulties to reach consensus among the Bootspec ecosystem
developers and maintainers, we should reach out to the standard RFC process.

# Goals
[goals]: #goals

- Improve non-x86 support in Bootspec, emphasizing the importance of
  devicetrees.
- Address and reduce the risks associated with initrd secrets, providing a

>   default "secure" implementation within the systemd ecosystem and offering

Author: My current feeling about scripted initrd is that we will start deprecation in 24.05, thus, getting the nice features for it seems to be a waste of energy, but I may be wrong (?).

We can keep v1 working as-is for scripted initrd as we go in deprecation of v1 too, no?

### Bootspec Format v2
[format-v2]: #format-v2

Using the following JSON:

```json5
{
  // Toplevel key describing the version of the specification used in the document
  "org.nixos.bootspec.v2": {
    // (Required) System type the bootspec is intended for (e.g. `x86_64-linux`, `aarch64-linux`)
    "system": "x86_64-linux",

    // (Required) Path to the stage-2 init, executed by the initrd (if present)
    "init": "/nix/store/xxx-nixos-system-xxx/init",

    // (Required) List of paths to the initrd, can be empty
    "initrds": [ "/nix/store/xxx-initrd-linux/initrd" ],

    // (Required) Path to the kernel image
    "kernel": "/nix/store/xxx-linux/bzImage",

    // (Required) Kernel commandline options
    "kernelParams": [
      "amd_iommu=on",
      "amd_iommu=pt",
      "iommu=pt",
      "kvm.ignore_msrs=1",
      "kvm.report_ignored_msrs=0",
      "udev.log_priority=3",
      "systemd.unified_cgroup_hierarchy=1",
      "loglevel=4"
    ],

    // (Required) The label of the system. It should contain the operating system, kernel version,
    // and other user-relevant information to identify the system. This corresponds
    // loosely to `config.system.nixos.label`.

>     "label": "NixOS 21.11.20210810.dirty (Linux 5.15.30)",

Reviewer 6: I'd like this to be structured information, i.e. make label an attrset that contains `distroName, `release`, `uname` etc. This is because ultimately it is the bootloader builder's resposibility to assemble this information so that it can be displayed nicely using the facilities the specific bootloader provdes. What is "nice" might also change between bootloaders.

We have this problem in Lanzaboote already where we have a suboptimal display name for a generation because we have to just use the label as is.

Reviewer 3: I believe this is answered here:

 - https://github.com/NixOS/rfcs/pull/165#issuecomment-1859188215

* * *

My understanding here is that providing structured data is undesirable as this represents a snapshot of the already serialized boot entry configuration. It is not intended to be re-parsed for further processing.

In this instance, it is the [`system.nixos.label`](https://search.nixos.org/options?channel=unstable&show=system.nixos.label&from=0&size=50&sort=alpha_asc&type=packages&query=++++system.nixos.label) configuration, which is an arbitrary label that can be configured by the end-user.

What, really, I'm wondering is: is the comment describing it correct? Shouldn't it be just `system.nixos.label`, and if not, why?

Reviewer 7: Generally the bootloader would use the generation number, except that the generation number is nowhere to be seen in the bootspec, and for good reason, since it's the one thing that's guaranteed to change between rebuilds, while the boot.json resides within the system path itself, which is only supposed to depend on what's explicitly provided to nix at build-time. Bootspec doesn't even provide _any_ information that other generations even exist beyond specialisations for that particular build.

So bootloaders are already forced to look elsewhere than just the provided system path in order to generate its boot entries. At that point, why not just take a peek at /etc/os-release to get more granular information and reduce some of the duplication? If the boot menu has any kind of hierarchy, then saying "NixOS" for every single subentry seems unnecessary when the version information is all you need. The only information in `label` that isn't in `os-release` is the kernel version.

Author: Any bootloader implementation can peek at the `/etc/os-release` as long as it's determined at build-time and generated an appropriate label.

Furthermore, there's no reason to make `/etc/os-release` a dependency, someone could use bootspec on systems without it and with a special bootloader implementation making use of another knowledge.

The kernel version is also known at build-time. There's no need to literally read `/etc/os-release` per se.

Ideally, the bootloader should try to honor as much as possible the `label` or come up with better generation of the `label`, sure, in lanzaboote, we have a problem that is, we need better tools to format the title because we can run in various excessive cleverness from systemd-boot that forces us to assume something about the label itself and manipulate it. As soon someone overrides forcibly that label, we are doomed.

But customization of boot title entry is a complicated matter, one, that exceeds the scope of bootspec maybe…

To put in another way, there's a tension between "bootspec's label is an opaque string that must be honored" and "my bootloader is doing weird things to the sorting order of the label / eating some words / I would like to show a 2-level menu and I don't know where to split the label string".

The resolution of this situation might be to have `preferredLabel` and the data @nikstur proposes to have which provides the preferred label in a structured way which the bootloader implementation may decide to reassemble for limitations of the software or to better accommodate the UI. Though, this increases the complexity of the design and opens a can of worms.

  // (Optional) Specialisations are an extension to the specification which allows bundling multiple variants of a NixOS configuration with a single parent.
  // These are shaped like the top level; to be precise:
  //  - Each entry in the toplevel "org.nixos.specialisation.v2" object represents a specialisation.
  //  - In order for the top-level document to be a valid v2 bootspec, each specialisation must have a valid "org.nixos.bootspec.v2" key whose value conforms to the same schema as the toplevel "org.nixos.bootspec.v2" object.
  //  - The behaviour of nested specialisations (i.e. entries in "org.nixos.specialisation.v2" which themselves contain the "org.nixos.specialisation.v2" key) is not defined.
  //  - In particular, there is no expectation that such nested specialisations will be handled by consumers of bootspec documents.
  //  - Each specialisation document may contain arbitrary further keys (extensions), like the top-level document.
  //  - The semantics of these should be the same as when these keys are used at the top level, but only apply for the given specialisation.

>   "org.nixos.specialisation.v2": {

Author: Well, each specialization must have a valid v2 bootspec document, so it's a change in the expectations of the v1 specialization specification, so it deserves a bump (?).

(I also like bumps to test fully our capabilities to avoid ossification though.)

    // Each key in this object corresponds to a specialisation as defined by the `specialisation.<name>` NixOS option.
    "<name>": {
      "org.nixos.bootspec.v2": {
        // See above
      }
    }
  },

  // (Optional) Hash map of desired secrets for that generation inside of the initrd.
  // Implementors of a bootloader installation procedure should examine their options
  // to securely make available the secret inside the initrd phase.
  // This may involve leveraging TPM2 via systemd-credentials or any measure you deem
  // to be reasonable in the context.

>   // The legacy behavior is to prepare a CPIO archive for each file and
>   // extend the `initrds` fields with those CPIO archives.

Reviewer 5: ```suggestion
  // The legacy behavior is to prepare a CPIO archive for each file and
  // concatenate these with the initrds from the `initrds` field to form
  // a single initrd for boot. This exposes the secrets to anyone who can
  // read the boot partition, and is discouraged in favour of mechanisms
  // which provide stronger protection for the secrets.
```

  // Make sure the location where the secrets are dropped in the initrd are visible

>   // for the user.

Reviewer 5: Unclear, what does this mean?

Author: Imagine you are activating the script in a rescue system and the secrets are in `/mnt/etc/secrets/...` instead of `/etc/secrets/...` and you are not fully chrooted inside `/mnt` or whatever, the location won't be visible at assembly time.

This sort of problem cause a lot of difficult to understand to most users because activation now depends on an invisible files to assemble bootables (arguably, this is not the only mechanism like this we have.).

How would you reframe this concept?
